src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2904,116 +2867,123 @@ private static bool IsReallyAType(TypeSymbol type)
+                var type = candidate;
+                if (!conversions.IncludeNullability)
                 {
-                    builder.AddRange(values);
+                    type = type.SetUnknownNullabilityForReferenceTypes();
+                }
+                if (!AddOrMergeCandidate(candidates, type, variance, conversions))
+                {
+                    return false;
                 }
             }
+            return true;
+        }
+
+        private static bool AddOrMergeCandidate(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            TypeSymbolWithAnnotations @new,
+            VarianceKind variance,
+            ConversionsBase conversions)
+        {
+            Debug.Assert(conversions.IncludeNullability ||
+                @new.SetUnknownNullabilityForReferenceTypes().Equals(@new, TypeCompareKind.CompareNullableModifiersForReferenceTypes));
 
-            internal int Count
+            if (candidates.TryGetValue(@new, out TypeSymbolWithAnnotations old))
             {
-                get { return _dictionary.Count; }
+                return TryMergeAndReplaceIfStillCandidate(candidates, old, @new,  variance, conversions);
             }
-
-            internal void AddAll(HashSet<TypeSymbolWithAnnotations> set, ConversionsBase conversions)
+            else
             {
-                foreach (var candidate in set)
-                {
-                    Add(candidate, conversions);
-                }
+                candidates.Add(@new, @new);
+                return true;
             }
+        }
 
-            internal void Add(TypeSymbolWithAnnotations type, ConversionsBase conversions)
+        private static void MergeOrRemoveCandidates(
+            Dictionary<TypeSymbolWithAnnotations, TypeSymbolWithAnnotations> candidates,
+            HashSet<TypeSymbolWithAnnotations> bounds,
+            ArrayBuilder<TypeSymbolWithAnnotations> initialCandidates,
+            ConversionsBase conversions,
+            VarianceKind variance,
+            ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            // SPEC:   For each lower (upper) bound U of Xi all types to which there is not an
+            // SPEC:   implicit conversion from (to) U are removed from the candidate set.
+            foreach (var bound in bounds)
             {
-                if (Merge(type, conversions))
-                {
-                    return;
-                }
-
-                List<TypeSymbolWithAnnotations> list;
-                if (!_dictionary.TryGetValue(type, out list))
+                foreach (var candidate in initialCandidates)
                 {
-                    list = new List<TypeSymbolWithAnnotations>();
-                    _dictionary.Add(type, list);
+                    if (bound.Equals(candidate, conversions.IncludeNullability ? TypeCompareKind.CompareNullableModifiersForReferenceTypes : TypeCompareKind.ConsiderEverything))
```

Comment:
>conversions.IncludeNullability ? TypeCompareKind.CompareNullableModifiersForReferenceTypes : TypeCompareKind.ConsiderEverything [](start = 48, length = 127)

Consider extracting out of the loops.
Same for `conversions.WithNullability(false)` below (line 2946)

Reply:
Left `conversions.WithNullability(false)` as is since it inexpensive to call the second time.

---
In reply to: [223845493](https://github.com/dotnet/roslyn/pull/29727#discussion_r223845493) [](ancestors = 223845493)